File: src/schedulerStore.ts
import { existsSync, readFileSync, writeFileSync, mkdirSync } from "fs";
import path from "path";
import { randomUUID } from "crypto";
import { resolveDataPath } from "./config.js";

export type ScheduledTask = {
  id: string;
  userId: number;
  lang: 'en' | 'es';
  locale: string;
  timeZone: string;
  scheduledTime: Date;
  createdAt: Date;
  executedAt?: Date;
  status: 'pending' | 'executed' | 'failed';
  error?: string;
}

type SerializedTask = Omit<ScheduledTask, 'scheduledTime' | 'createdAt' | 'executedAt'> & {
  scheduledTime: string;
  createdAt: string;
  executedAt?: string;
};

export class SchedulerStore {
  schedules: ScheduledTask[] = [];
  private filePath: string;

  constructor(filePath = resolveDataPath("scheduledTasks.json")) {
    this.filePath = filePath;
    const dir = path.dirname(this.filePath);
    mkdirSync(dir, { recursive: true });
    this.loadFromDisk();
  }

  private toSerializable(task: ScheduledTask): SerializedTask {
    return {
      ...task,
      scheduledTime: task.scheduledTime.toISOString(),
      createdAt: task.createdAt.toISOString(),
      executedAt: task.executedAt ? task.executedAt.toISOString() : undefined,
    };
  }

  private fromSerializable(task: SerializedTask): ScheduledTask {
    return {
      ...task,
      lang: task.lang ?? 'en',
      locale: task.locale ?? 'en',
      timeZone: task.timeZone ?? 'Europe/Madrid',
      scheduledTime: new Date(task.scheduledTime),
      createdAt: new Date(task.createdAt),
      executedAt: task.executedAt ? new Date(task.executedAt) : undefined,
    };
  }

  add(
    userId: number,
    scheduledTime: Date,
    lang: 'en' | 'es' = 'en',
    locale: string = 'en',
    timeZone: string = 'Europe/Madrid'
  ): ScheduledTask {
    const task: ScheduledTask = {
      id: randomUUID(),
      userId,
      lang,
      locale,
      timeZone,
      scheduledTime,
      createdAt: new Date(),
      status: 'pending',
    };

    this.schedules.push(task);
    this.saveToDisk();
    return task;
  }

  getPending(): ScheduledTask[] {
    return this.schedules
      .filter((task) => task.status === 'pending')
      .sort((a, b) => a.scheduledTime.getTime() - b.scheduledTime.getTime());
  }

  getByUser(userId: number): ScheduledTask[] {
    return this.schedules
      .filter((task) => task.userId === userId)
      .sort((a, b) => a.scheduledTime.getTime() - b.scheduledTime.getTime());
  }

  markExecuted(id: string, error?: Error | string): boolean {
    const task = this.schedules.find((item) => item.id === id);
    if (!task) return false;

    if (error) {
      task.status = 'failed';
      task.error = typeof error === 'string' ? error : error.message;
    } else {
      task.status = 'executed';
      task.executedAt = new Date();
    }

    this.saveToDisk();
    return true;
  }

  cancel(id: string): boolean {
    const task = this.schedules.find((item) => item.id === id);
    if (!task) return false;

    task.status = 'failed';
    task.error = 'cancelled';
    this.saveToDisk();
    return true;
  }

  saveToDisk() {
    try {
      const serialized = this.schedules.map((task) => this.toSerializable(task));
      writeFileSync(this.filePath, JSON.stringify(serialized, null, 2), "utf8");
    } catch (error) {
      console.log(`error writing file ${error}`);
    }
  }

  loadFromDisk() {
    try {
      if (!existsSync(this.filePath)) {
        this.schedules = [];
        return;
      }

      const data = readFileSync(this.filePath, "utf8");
      const parsed = JSON.parse(data) as SerializedTask[];
      this.schedules = parsed.map((task) => this.fromSerializable(task));
    } catch (error) {
      console.log(`error loading file ${error}`);
      this.schedules = [];
    }
  }
}

File: rust-src/scheduler_store.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::Result;
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::types::{Lang, ScheduledTask, TaskStatus};

#[derive(Debug)]
pub struct SchedulerStore {
    file_path: PathBuf,
    schedules: Vec<ScheduledTask>,
}

impl SchedulerStore {
    pub fn new(path: impl AsRef<Path>) -> Result<Self> {
        let file_path = path.as_ref().to_path_buf();
        if let Some(parent) = file_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut store = Self {
            file_path,
            schedules: Vec::new(),
        };
        store.load_from_disk();
        Ok(store)
    }

    pub fn add(
        &mut self,
        user_id: i64,
        scheduled_time: DateTime<Utc>,
        lang: Lang,
        locale: String,
        time_zone: String,
    ) -> ScheduledTask {
        let task = ScheduledTask {
            id: Uuid::new_v4().to_string(),
            user_id,
            lang,
            locale,
            time_zone,
            scheduled_time,
            created_at: Utc::now(),
            executed_at: None,
            status: TaskStatus::Pending,
            error: None,
        };

        self.schedules.push(task.clone());
        let _ = self.save_to_disk();
        task
    }

    pub fn get_pending(&self) -> Vec<ScheduledTask> {
        let mut tasks: Vec<_> = self
            .schedules
            .iter()
            .filter(|t| t.status == TaskStatus::Pending)
            .cloned()
            .collect();
        tasks.sort_by_key(|t| t.scheduled_time);
        tasks
    }

    pub fn get_by_user(&self, user_id: i64) -> Vec<ScheduledTask> {
        let mut tasks: Vec<_> = self
            .schedules
            .iter()
            .filter(|t| t.user_id == user_id)
            .cloned()
            .collect();
        tasks.sort_by_key(|t| t.scheduled_time);
        tasks
    }

    pub fn mark_executed(&mut self, id: &str, error: Option<String>) -> bool {
        let Some(task) = self.schedules.iter_mut().find(|t| t.id == id) else {
            return false;
        };

        if let Some(err) = error {
            task.status = TaskStatus::Failed;
            task.error = Some(err);
        } else {
            task.status = TaskStatus::Executed;
            task.executed_at = Some(Utc::now());
            task.error = None;
        }

        let _ = self.save_to_disk();
        true
    }

    pub fn cancel(&mut self, id: &str) -> bool {
        let Some(task) = self.schedules.iter_mut().find(|t| t.id == id) else {
            return false;
        };

        task.status = TaskStatus::Failed;
        task.error = Some("cancelled".to_string());
        let _ = self.save_to_disk();
        true
    }

    fn load_from_disk(&mut self) {
        if !self.file_path.exists() {
            self.schedules = Vec::new();
            return;
        }

        match fs::read_to_string(&self.file_path) {
            Ok(content) => match serde_json::from_str::<Vec<ScheduledTask>>(&content) {
                Ok(parsed) => self.schedules = parsed,
                Err(err) => {
                    eprintln!("error loading schedule json: {err}");
                    self.schedules = Vec::new();
                }
            },
            Err(err) => {
                eprintln!("error reading schedule file: {err}");
                self.schedules = Vec::new();
            }
        }
    }

    fn save_to_disk(&self) -> Result<()> {
        let json = serde_json::to_string_pretty(&self.schedules)?;
        fs::write(&self.file_path, json)?;
        Ok(())
    }
}
